core/genesis.go

```diff
@@ -246,21 +246,115 @@ func SetupGenesisBlockWithOverride(db ethdb.Database, genesis *Genesis, override
 	return newcfg, stored, nil
 }
 
+// Hard fork block height specified in config.toml has higher priority, but
+// if it is not specified in config.toml, use the default height in code.
 func (g *Genesis) configOrDefault(ghash common.Hash) *params.ChainConfig {
+	var defaultConfig *params.ChainConfig
 	switch {
-	case g != nil:
-		return g.Config
 	case ghash == params.MainnetGenesisHash:
-		return params.MainnetChainConfig
+		defaultConfig = params.MainnetChainConfig
 	case ghash == params.BSCGenesisHash:
-		return params.BSCChainConfig
+		defaultConfig = params.BSCChainConfig
 	case ghash == params.ChapelGenesisHash:
-		return params.ChapelChainConfig
+		defaultConfig = params.ChapelChainConfig
 	case ghash == params.RialtoGenesisHash:
-		return params.RialtoChainConfig
+		defaultConfig = params.RialtoChainConfig
 	default:
-		return params.AllEthashProtocolChanges
+		defaultConfig = params.AllEthashProtocolChanges
 	}
+	if g == nil || g.Config == nil {
+		return defaultConfig
+	}
+
+	// if not set in config.toml, use the default value of each network
+	if g.Config.ChainID == nil {
+		g.Config.ChainID = defaultConfig.ChainID
+	}
+	if g.Config.HomesteadBlock == nil {
+		g.Config.HomesteadBlock = defaultConfig.HomesteadBlock
+	}
+	if g.Config.EIP150Block == nil {
+		g.Config.EIP150Block = defaultConfig.EIP150Block
+	}
+	if g.Config.EIP155Block == nil {
+		g.Config.EIP155Block = defaultConfig.EIP155Block
+	}
+	if g.Config.EIP158Block == nil {
+		g.Config.EIP158Block = defaultConfig.EIP158Block
+	}
+	if g.Config.ByzantiumBlock == nil {
+		g.Config.ByzantiumBlock = defaultConfig.ByzantiumBlock
```

Comment:
Just curious how could we achieve the goal of automatically update? , it seems that  by using reflect.VisibleFields, it still need to use 'if' to compare the fieldName and check its value, and then set the default value (BlockNumber) if nil.

Another minor comments is that reflection seems to be a little hard to maintain as it kind of `wrap` the type of the field when trying to access it's value, and also the rule of `settable`. which imbo becomes error-prone for developers when something new is added to `ChainConfig`.

Reply:
@sunny2022da We can compare field names by converting them to string using reflection. Then for each field name of `g.Config`  we look up the value of that field in `defaultConfig` using the string as a key. Then through reflection we can set the value of the field in `g.Config`.

This is actually how geth maps RPC method names like `eth_gasPrice` to callbacks such as `func GasPrice(..)`.

Reply:
> 

got you~  So it seems to me the logic is:

- use reflect.visibleFields to get the list of all `*Block' fields
- iterate every `field = fields[i]` , compare the fieldName and the name of `HF block`, such as `HomesteadBlock `, then get the value of the field.
- check if field value is nil, if yes, assign to the defaultConfig value. 

it seems to me that the logic is a little bit complicated then the one we currently have.

Moreover, I think the reason that geth use reflection for mapping RPC calls is because there's serialization/de-serialization involved in the RPC case, where reflection is a good choice to get the method name.

Reply:
You are more or less right. You can do it all in one for loop. You iterate over the fields of `g.Config`, and if the value is nil, you retrieve using the map structure the matching block in `defaultConfig`, and update the field's value. 

If you use this map you don't need to add an if statement for each block name since the key of the map will find you the matching block from `defaultConfig`.

I think this logic is not that complicated, and even if the code ends up being complicated, the user experience and maintainability are much better as we wouldn't need to keep updating this code every time there is a new hard fork.

Reply:
Correct, we just need to maintain the map you mentioned in this way, addint new <k,v> pair when there is new HF added. 
So the problem remain to whether we decide to use reflection here, Let hear others~ 

Thx for explanation!

Reply:
So this PR should add default config values if not specified in config.toml. In this case why reinventing the wheel and not use a config package which is already tested?
ex 1: https://github.com/gookit/config
ex 2: https://pkg.go.dev/github.com/ardanlabs/conf/v3#hdr-Example_Usage

Reply:
Thank for all your comments, I think the main point is whether to improve the code by reflect.
I would prefer the if/else logic, since:
1.could avoid the complexity of reflection
2.the extra KV map mentioned by @Mister-EA could bring big change to the current HF setup.